typescript/index.d.ts

```diff
@@ -88,22 +88,30 @@ export interface FieldProps<T extends HTMLElement>
   [otherProp: string]: any;
 }
 
-export interface UseFormStateParams {
-  onChange?: (formState: FormState) => void;
+export interface UseFormStateParams<FormValues = object> {
+  onChange?: (formState: FormState<FormValues>) => void;
   subscription?: FormSubscription;
 }
 
-export interface FormSpyProps
-  extends UseFormStateParams,
-    RenderableProps<FormSpyRenderProps> {}
+export interface FormSpyProps<FormValues>
+  extends UseFormStateParams<FormValues>,
+    RenderableProps<FormSpyRenderProps<FormValues>> {}
 
 export const Field: React.FC<FieldProps<any>>;
 export const Form: React.FC<FormProps<object>>;
```

Comment:
I was actually about to submit a PR very similar to this. I just implemented final-form at the place I work and love it. I just wish the types were a bit stronger. While this PR is very close to what I had the `Form` component is still just getting the `object` generic type passed to it.

I actually implemented something like this:
```
export const Form: <FormValues = object>(
  props: FormProps<FormValues>
) => React.FunctionComponentElement<FormProps<FormValues>>;
```
This should allow typescript to infer the types based on the `initialValues` props passed to the form or you could set the generic to the form itself like so:

`<Form<{foo: string, bar: number }> initialValues={initialValues} onSubmit={onSubmit}>`

This similar to what is in #426, but I'm not using a class since it isn't one. 

I also did the same to `Field`:
```
export const Field: <FieldValue = any, T extends HTMLElement = HTMLElement>(
  props: FieldProps<FieldValue, T>
) => React.FunctionComponentElement<FieldProps<FieldValue, T>>;
```

This way I could pass the type `FieldType` to `UseFieldConfig` and then the `initalValue` prop and others on `Field` would be strongly typed:

`<Field<string> name="foo" label="Foo" initialValue="bar" component={TextField} required validate={required} />`

I also did the same thing for field arrays. So when you were mapping through the array and needed to get the value off an index, it would have a type.

Reply:
That's very interesting.

I had no idea that you could do `<Field<string> name="foo">` in TSX. That's wild.

Reply:
Yea, I've used it a couple of times. I think it's been in typescript since version 2.9 or so. Just allowing for a generic like that could allow something like the Form `component` to infer types sometimes too.

I'd be happy to try and implement the solution above after this PR would get merged. I think at least having generic typing for the `Form` component would be very nice instead of it just being set to `object`.